File: space_sketcher/rna/src/dbscan_filter.R
##This script is used to perform dbscan clustering for each cell barcode.
##And then calculate the ratio of single-cluster, multi-cluster and no-cluster.
##In single-cluster, calculate the weighted centroid of the cluster.
##In multi-cluster, choose the max cluster as the cluster, and calculate the weighted centroid of the cluster.
##Filter out no-cluster cell barcodes.
library(ggplot2)
library(dplyr)
library(readr)
library(dbscan)
library(gridExtra)
library(argparse)


# Define the command line arguments
parser <- ArgumentParser()
parser$add_argument("--infile", type = "character", help = "The directory of the matrix")
parser$add_argument("--outdir", type = "character", help = "The output directory")
parser$add_argument("--maxumi", type = "numeric", help = "The number of max umi count of spatial barcode to filter out, default: 5000", default = 5000)
parser$add_argument("--minumi", type = "numeric", help = "The number of min umi count of spatial barcode to filter out, default: 2", default = 2)
args <- parser$parse_args()

infile <- args$infile
samplename <- args$samplename
outdir <- args$outdir
maxumi <- args$maxumi
minumi <- args$minumi

rm(list=ls())
# infile <- "/data01/huanghuichang/spatialRNA_analysis/20250120/N16-puck-9-test/N16-puck-9.CB_coord.total.txt"
# samplename <- "N16-puck-9-filterumi-lowandhigh"
# outdir <- "/data01/home/huanghuichang/Test/Space-Sketcher/Space-sketcher/space_sketcher/rna/src"

#####>>>>>Load data<<<<<#####
CB_coord_total <- read_delim(infile, delim = "\t", escape_double = FALSE, trim_ws = TRUE)
head(CB_coord_total)

####total umi of each spatial barcode
subsb_umi_summary <- CB_coord_total %>%
  group_by(subsb) %>%
  summarise(umi_count_sum = sum(umi_count))
###sort
subsb_umi_summary <- subsb_umi_summary %>%
  arrange(desc(umi_count_sum))
###add rank
subsb_umi_summary <- subsb_umi_summary %>%
  mutate(rank = row_number())
head(subsb_umi_summary)

### 筛选出umi_count_sum大于maxumi或小于minumi的subsb
subsb_to_filter <- subsb_umi_summary %>%
  filter(umi_count_sum > maxumi | umi_count_sum < minumi) %>%
  select(subsb)  # 仅保留subsb列

CB_coord_total_filtered <- CB_coord_total %>%
  filter(!subsb %in% subsb_to_filter$subsb)

#####>>>>>DBSCAN cluster<<<<<#####
###unique cell barcodes
unique_barcodes <- unique(CB_coord_total_filtered$cb)
no_cluster <- 0
single_cluster <- 0
multi_cluster <- 0
filtereddf <- data.frame()
cb_cluster <- data.frame()
###clustering for each cell barcode
for (barcode in unique_barcodes) {

  cb_sel <- CB_coord_total_filtered %>% filter(cb == barcode) %>% arrange(umi_count)
  ###check if there are enough data points to perform clustering
  if (nrow(cb_sel) < 6) {
    cb_sel$db_clu <- as.factor(0)
    no_cluster <- no_cluster + 1
    clustertype <- "no-cluster"
    cb_cluster_temp <- data.frame(cb = barcode, cluster = clustertype)
    cb_cluster <- rbind(cb_cluster, cb_cluster_temp)
    next
  }

  db <- dbscan(cb_sel[, c("xcoord", "ycoord")], eps = 150, minPts = 6, weights = sqrt(cb_sel$umi_count))
  cb_sel$db_clu <- as.factor(db$cluster)
  ######detect multi-clusters or single-cluster
  head(cb_sel)
  cluster_table <- table(cb_sel$db_clu)
  cluster_counts <- as.numeric(cluster_table)
  ###remove the first cluster count cause it is the noise cluster
  cluster_counts <- cluster_counts[-1]
  if (length(cluster_counts) > 0) {
    maxcluser <- max(cluster_counts)
    if (length(cluster_counts) > 1) {
      second_max <- sort(cluster_counts, decreasing = TRUE)[2]
    } else {
      second_max <- 0
    }
    if (second_max > maxcluser/2) {
      multi_cluster <- multi_cluster + 1
      clustertype <- "multi-cluster"
    } else {
      single_cluster <- single_cluster + 1
      clustertype <- "single-cluster"
    }
    ###Weighted centroid calculation
    cluster1_cell_cb <- cb_sel %>% filter(db_clu == which.max(cluster_counts))
    xcor <- cluster1_cell_cb$xcoord
    ycor <- cluster1_cell_cb$ycoord
    weights <- cluster1_cell_cb$umi_count
    weighted_centroid_x <- round(sum(xcor * weights) / sum(weights),0)
    weighted_centroid_y <- round(sum(ycor * weights) / sum(weights),0)
    weighted_centroid_cell <- data.frame(cb = barcode, xcoord = weighted_centroid_x, ycoord = weighted_centroid_y)
    filtereddf <- rbind(filtereddf, weighted_centroid_cell)
  } else {
    no_cluster <- no_cluster + 1
    clustertype <- "no-cluster"
  }
  cb_cluster_temp <- data.frame(cb = barcode, cluster = clustertype)
  cb_cluster <- rbind(cb_cluster, cb_cluster_temp)
}

###calculating the ratio of single-cluster, multi-cluster and no-cluster
cluster_type <- c("single_cluster", "multi_cluster", "no_cluster")
cluster_counts <- c(single_cluster, multi_cluster, no_cluster)
cluster_ratio <- round(cluster_counts*100/sum(cluster_counts), 2)
clusterdf <- data.frame(cluster_type = cluster_type, counts = cluster_counts, ratio = cluster_ratio)

output_file <- paste0(outdir, "/dbscan_cluster_distribution.csv")
write.table(clusterdf, file = output_file, sep = ",", row.names = FALSE, col.names = TRUE, quote = FALSE)


head(filtereddf)
outfile <- paste0(outdir, "/CB_coord.dbscan_filtered.txt")
write.table(filtereddf, file = outfile, sep = "\t", row.names = FALSE, col.names = TRUE, quote = FALSE)

###unique cell barcodes
cb_cluster <- unique(cb_cluster)
outcbcluster <- paste0(outdir, "/CB_cluster.txt")
write.table(cb_cluster, file = outcbcluster, sep = "\t", row.names = FALSE, col.names = TRUE, quote = FALSE)

#####>>>>>Plot data statistic<<<<<#####
##spatial barcode - umi 骤降图
p <- ggplot(subsb_umi_summary, aes(x = rank, y = umi_count_sum)) +
  geom_point() +  # 绘制散点图
  geom_line() +   # 绘制连接线
  geom_hline(yintercept = 5000, linetype = "dashed", color = "red") +  
  geom_hline(yintercept = 2, linetype = "dashed", color = "blue") +  
  scale_x_log10() +  # 设置x轴为对数刻度
  scale_y_log10() +  # 设置y轴为对数刻度
  labs(
    title = "UMI Count Sum of Each Spatial Barcode",
    x = "Rank",
    y = "UMI Count Sum"
  ) +
  theme_minimal() +  # 使用简洁主题
  theme(
    panel.background = element_rect(fill = "white"),  # 设置面板背景为白色
    plot.background = element_rect(fill = "white", color = "white")  # 设置整个图形背景为白色
  )

ggsave(paste0(outdir,"/spatial_umi_knee_plot.png"), plot = p, width = 8, height = 6, dpi = 300)
###output subsb_umi_summary
subsb_umi_summary_out <- paste0(outdir,"/spatial_umi_knee_plot.temp.txt")
write.table(subsb_umi_summary, file = subsb_umi_summary_out, sep = "\t", row.names = FALSE, col.names = TRUE, quote = FALSE)


##cell barcode - mean umi count of top 100 spatial barcode 骤降图
# 将clustertype中的cluster信息合并到CB_coord_total_filtered中
CB_coord_total_filtered <- CB_coord_total_filtered %>%
  left_join(cb_cluster, by = "cb")

# 按cb分组计算每个细胞中top100 sb的umi count 的均值
cb_umi_mean_top100 <- CB_coord_total_filtered %>%
  group_by(cb) %>%
  arrange(desc(umi_count)) %>%  # 按umi_count降序排序
  slice_head(n = 100) %>%  # 选择每个cb组的前100个sb
  group_by(cb, cluster) %>%  # 重新按cb分组
  summarise(umi_count_mean = mean(umi_count)) %>%  # 计算前100个sb的umi_count均值
  ungroup()  # 去掉分组

# 按umi_count_mean降序排序
cb_umi_mean_top100 <- cb_umi_mean_top100 %>%
  arrange(desc(umi_count_mean))
# 添加序号列
cb_umi_mean_top100 <- cb_umi_mean_top100 %>%
  mutate(rank = row_number())

median_top100_umi_cell_mean <- median(cb_umi_mean_top100$umi_count_mean)
mean_top100_umi_cell_mean <- mean(cb_umi_mean_top100$umi_count_mean)

p <- ggplot(cb_umi_mean_top100, aes(x = rank, y = umi_count_mean, color = cluster)) +
  geom_point() +  # 绘制散点图
  geom_line() +   # 绘制连接线
  scale_x_log10() +  # 设置x轴为对数刻度
  scale_y_log10() +  # 设置y轴为对数刻度
  labs(
    title = "Mean of UMI Count of Top 100 Spatial Barcodes in Each Cell",
    x = "Rank",
    y = "Mean of UMI count",
    color = "Cluster"
  ) +
  theme_minimal() +  # 使用简洁主题
  theme(
    panel.background = element_rect(fill = "white"),  # 设置面板背景为白色
    plot.background = element_rect(fill = "white", color = "white")  # 设置整个图形背景为白色
  )

ggsave(paste0(outdir, "/cb_umi_mean_knee_plot.png"), plot = p, width = 8, height = 6, dpi = 300)
###output cb_umi_mean_top100
cb_umi_mean_top100_out <- paste0(outdir,"/cb_umi_mean_knee_plot.temp.txt")
write.table(cb_umi_mean_top100, file = cb_umi_mean_top100_out, sep = "\t", row.names = FALSE, col.names = TRUE, quote = FALSE)


###output new stat
output_file <- file(paste0(outdir,"/.sb_cb_umi_stats.csv"), "w")
writeLines(paste("median_top100_umi_cell_mean,", median_top100_umi_cell_mean), output_file)
writeLines(paste("mean_top100_umi_cell_mean,", mean_top100_umi_cell_mean), output_file)


###cell barcode vs spatial barcodes 种类数骤降图
# 将clustertype中的cluster信息合并到CB_coord_total中
CB_coord_total <- CB_coord_total %>%
  left_join(cb_cluster, by = "cb")

sb_count <- CB_coord_total %>%
  group_by(cb,cluster) %>%
  summarise(sbcounts = n_distinct(sb)) %>%
  ungroup()

# 根据种类数量从大到小排序，并新增序号列
sb_count <- sb_count %>%
  arrange(desc(sbcounts)) %>%
  mutate(rank = row_number())

# 绘制对数-对数图
p <- ggplot(sb_count, aes(x = rank, y = sbcounts, color = cluster)) +
  geom_point() +  # 绘制散点图
  geom_line() +   # 绘制连接线
  scale_x_log10() +  # 设置x轴为对数刻度
  scale_y_log10() +  # 设置y轴为对数刻度
  labs(
    title = "Spatial Barcode Counts of Each Cell",
    x = "Rank",
    y = "sb counts"
  ) +
  theme_minimal() +  # 使用简洁主题
  theme(
    panel.background = element_rect(fill = "white"),  # 设置面板背景为白色
    plot.background = element_rect(fill = "white", color = "white")  # 设置整个图形背景为白色
  )

ggsave(paste0(outdir, "/cb_sb_counts.png"), plot = p, width = 8, height = 6, dpi = 300)

File: space_sketcher/rna/src/dbscan_filter.py
#!/usr/bin/env python3
import os
import argparse
import pandas as pd
import numpy as np
from sklearn.cluster import DBSCAN
from joblib import Parallel, delayed
from tqdm import tqdm  # 进度条支持
import scanpy as sc
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.sparse import csr_matrix
import scipy.io


def read_data(infile):
    """Read input file and return DataFrame"""
    return pd.read_csv(infile, sep='\t')

def write_10x_format(outdir, adata, gene_symbols=True):
    """
    输出10X Genomics标准格式（Seurat兼容）
    
    参数:
        outdir: 输出目录路径
        adata: AnnData对象
        gene_symbols: 是否使用基因符号作为feature名称
    """
    os.makedirs(outdir, exist_ok=True)
    
    # 1. 写入barcodes.tsv
    barcodes_path = os.path.join(outdir, "barcodes.tsv")
    pd.DataFrame(adata.obs_names).to_csv(
        barcodes_path, 
        sep='\t', 
        header=False, 
        index=False
    )
    
    # 2. 写入features.tsv
    features_path = os.path.join(outdir, "features.tsv")
    if gene_symbols and 'gene_symbols' in adata.var.columns:
        features_df = pd.DataFrame({
            'gene_ids': adata.var.index,
            'gene_symbols': adata.var['gene_symbols'],
            'feature_type': 'Gene Expression'
        })
    else:
        features_df = pd.DataFrame({
            'gene_ids': adata.var.index,
            'gene_symbols': adata.var.index,  # 如果没有gene_symbols则用gene_id
            'feature_type': 'Gene Expression'
        })
    features_df.to_csv(
        features_path, 
        sep='\t', 
        header=False, 
        index=False
    )
    
    # 3. 写入matrix.mtx (稀疏矩阵格式)
    matrix_path = os.path.join(outdir, "matrix.mtx")
    scipy.io.mmwrite(
        matrix_path,
        csr_matrix(adata.X.T)  # 注意需要转置
    )
    
    print(f"10X format files saved to: {outdir}")

def filter_by_umi(df, maxumi=5000, minumi=2):
    """Filter spatial barcodes by total UMI counts"""
    subsb_umi_summary = df.groupby('subsb')['umi_count'].sum().reset_index(name='umi_count_sum')
    subsb_umi_summary = subsb_umi_summary.sort_values('umi_count_sum', ascending=False)
    subsb_umi_summary['rank'] = range(1, len(subsb_umi_summary)+1)
    
    subsb_to_filter = subsb_umi_summary[
        (subsb_umi_summary['umi_count_sum'] > maxumi) | 
        (subsb_umi_summary['umi_count_sum'] < minumi)
    ]['subsb']
    
    return df[~df['subsb'].isin(subsb_to_filter)], subsb_umi_summary

def process_single_cell(barcode, cell_data, eps=150, min_samples=6):
    """
    处理单个细胞barcode的DBSCAN聚类
    参数:
        barcode: 当前处理的细胞barcode
        cell_data: 该细胞的所有空间坐标和UMI数据
        eps: DBSCAN半径参数
        min_samples: DBSCAN最小样本数
    返回:
        (barcode, centroid_coords, cluster_type)
    """
    # 数据准备
    coords = cell_data[['xcoord', 'ycoord']].values
    weights = np.sqrt(cell_data['umi_count'].values)
    
    # 跳过太小的簇
    if len(coords) < min_samples:
        return (barcode, None, "no-cluster")
    
    # DBSCAN聚类
    db = DBSCAN(
        eps=eps,
        min_samples=min_samples,
        metric='euclidean',
        algorithm='kd_tree',  # 更快的算法
        n_jobs=1  # 每个任务单线程
    ).fit(coords, sample_weight=weights)
    
    labels = db.labels_
    valid_labels = labels[labels != -1]  # 去除噪声点
    
    # 无有效簇的情况
    if len(valid_labels) == 0:
        return (barcode, None, "no-cluster")
    
    # 统计簇信息
    unique_labels, counts = np.unique(valid_labels, return_counts=True)
    main_cluster = unique_labels[np.argmax(counts)]
    main_mask = (labels == main_cluster)
    
    # 计算加权质心
    x_coords = cell_data['xcoord'].values[main_mask]
    y_coords = cell_data['ycoord'].values[main_mask]
    cluster_weights = cell_data['umi_count'].values[main_mask]
    
    centroid_x = round(np.sum(x_coords * cluster_weights) / np.sum(cluster_weights))
    centroid_y = round(np.sum(y_coords * cluster_weights) / np.sum(cluster_weights))
    
    # 判断簇类型
    if len(counts) > 1 and np.sort(counts)[-2] > max(counts)/2:
        cluster_type = "multi-cluster"
    else:
        cluster_type = "single-cluster"
    
    return (barcode, (centroid_x, centroid_y), cluster_type)

def perform_dbscan_parallel(df, eps=150, min_samples=6, n_jobs=4, batch_size=100):
    """
    并行DBSCAN处理
    参数:
        df: 包含所有细胞数据的DataFrame
        eps: DBSCAN半径参数
        min_samples: DBSCAN最小样本数 
        n_jobs: 并行任务数(-1表示使用所有核心)
        batch_size: 每个任务处理的细胞数
    返回:
        filtered_df, cb_cluster, cluster_stats
    """
    # 预处理：按细胞barcode分组
    grouped = df.groupby('cb')
    barcodes = list(grouped.groups.keys())
    
    # 并行处理
    results = Parallel(n_jobs=n_jobs, batch_size=batch_size)(
        delayed(process_single_cell)(barcode, grouped.get_group(barcode), eps, min_samples)
        for barcode in tqdm(barcodes, desc="Processing cells")
    )
    
    ## 有效细胞(有簇的)
    valid_results = [r for r in results if r[1] is not None]
    coord_data = {
        'cb': [r[0] for r in valid_results],
        'xcoord': [r[1][0] for r in valid_results],
        'ycoord': [r[1][1] for r in valid_results]
    }
    coord_df = pd.DataFrame(coord_data)
    
    ## 所有细胞的聚类类型
    cluster_data = {
        'cb': [r[0] for r in results],
        'cluster': [r[2] for r in results]
    }
    cb_cluster = pd.DataFrame(cluster_data)
    
    ## 聚类统计
    cluster_stats = cb_cluster['cluster'].value_counts().reset_index()
    cluster_stats.columns = ['cluster_type', 'counts']
    cluster_stats['ratio'] = np.round(
        cluster_stats['counts'] * 100 / cluster_stats['counts'].sum(), 
        2
    )
    
    return coord_df, cb_cluster, cluster_stats

def generate_plots(df, cb_cluster, subsb_umi_summary, outdir):
    """Generate diagnostic plots"""
    # Spatial barcode UMI knee plot
    plt.figure(figsize=(8, 6))
    plt.plot(subsb_umi_summary['rank'], subsb_umi_summary['umi_count_sum'], 'o-', markersize=3)
    plt.axhline(y=5000, color='red', linestyle='--')
    plt.axhline(y=2, color='blue', linestyle='--')
    plt.xscale('log')
    plt.yscale('log')
    plt.title("UMI Count Sum of Each Spatial Barcode")
    plt.xlabel("Rank")
    plt.ylabel("UMI Count Sum")
    plt.savefig(os.path.join(outdir, "spatial_umi_knee_plot.png"), dpi=300, bbox_inches='tight')
    plt.close()
    
    # Cell barcode mean UMI plot
    df = df.merge(cb_cluster, on='cb', how='left')
    
    cb_umi_mean_top100 = df.groupby('cb').apply(
        lambda x: x.nlargest(100, 'umi_count')['umi_count'].mean()
    ).reset_index(name='umi_count_mean')
    cb_umi_mean_top100 = cb_umi_mean_top100.merge(cb_cluster, on='cb', how='left')
    cb_umi_mean_top100 = cb_umi_mean_top100.sort_values('umi_count_mean', ascending=False)
    cb_umi_mean_top100['rank'] = range(1, len(cb_umi_mean_top100)+1)
    

    colordict = {"single-cluster":"lightskyblue",
                 "multi-cluster": "lightsalmon",
                 "no-cluster": "limegreen"}
    
    plt.figure(figsize=(8, 6))
    sns.scatterplot(data=cb_umi_mean_top100, 
                    x='rank', y='umi_count_mean', 
                    hue='cluster', linewidth = 0,
                    palette=colordict,
                    legend=False)
    sns.lineplot(data=cb_umi_mean_top100, 
                    x='rank', y='umi_count_mean', 
                    hue='cluster', markers=True,
                    linewidth=2,
                    palette=colordict)
    plt.xscale('log')
    plt.yscale('log')
    plt.title("Mean of UMI Count of Top 100 Spatial Barcodes in Each Cell")
    plt.xlabel("Rank")
    plt.ylabel("Mean of UMI count")
    plt.savefig(os.path.join(outdir, "cb_umi_mean_knee_plot.png"), dpi=300, bbox_inches='tight')
    plt.close()
    
    # Cell barcode vs spatial barcodes count plot
    sb_count = df.groupby(['cb', 'cluster'])['sb'].nunique().reset_index(name='sbcounts')
    sb_count = sb_count.sort_values('sbcounts', ascending=False)
    sb_count['rank'] = range(1, len(sb_count)+1)
    
    plt.figure(figsize=(8, 6))
    sns.scatterplot(data=sb_count, 
                    x='rank', y='sbcounts', 
                    hue='cluster', linewidth = 0,
                    palette=colordict,
                    legend=False)
    sns.lineplot(data=sb_count, 
                    x='rank', y='sbcounts', 
                    hue='cluster', markers=True,
                    linewidth=2,
                    palette=colordict)
    plt.xscale('log')
    plt.yscale('log')
    plt.title("Spatial Barcode Counts of Each Cell")
    plt.xlabel("Rank")
    plt.ylabel("sb counts")
    plt.savefig(os.path.join(outdir, "cb_sb_counts.png"), dpi=300, bbox_inches='tight')
    plt.close()
    
    return cb_umi_mean_top100

def filter_matrix(matrixdir, coord_df, library, outdir):
    """Filter matrix based on DBSCAN results"""
    adata = sc.read_10x_mtx(matrixdir, var_names='gene_symbols', cache=True)
    
    if library == "leader_v1":
        coord_df['cb'] = coord_df['cb'].apply(lambda x: f"{x[:10]}_{x[10:20]}")
    
    adata_barcodes = adata.obs_names.values
    cell_barcodes = coord_df['cb'].tolist()
    # 验证barcode交集
    valid_barcodes = set(cell_barcodes) & set(adata_barcodes)
    print(f"Found {len(valid_barcodes)} matching barcodes (originally {len(cell_barcodes)})")

    # 创建安全的布尔索引
    mask = np.isin(adata_barcodes, list(valid_barcodes))
    # 安全过滤
    filtered_adata = adata[mask, :].copy()
    # 输出前检查
    assert len(filtered_adata) == len(valid_barcodes), "过滤后维度不匹配"    

    outmtxdir = os.path.join(outdir, "SCST")
    if os.path.exists(outmtxdir):
        import shutil
        shutil.rmtree(outmtxdir)
        
    write_10x_format(outmtxdir, filtered_adata)
    coord_df.to_csv(os.path.join(outmtxdir, "spatial_location_information.txt"), sep='\t', index=False)

    
def generate_summary(cellreads, coord_df, cb_umi_mean_top100, outdir):
    """Generate summary statistics"""
    cellreadsdata = pd.read_csv(cellreads, sep='\t')
    cellreadsdata = cellreadsdata[cellreadsdata['CB'] != "CBnotInPasslist"]
    
    if 'library' in locals() and 'library' == "leader_v1":
        cell_barcodes = coord_df['cb'].tolist()
    else:
        cell_barcodes = coord_df['cb'].tolist()
    
    truecellreads = cellreadsdata[cellreadsdata['CB'].isin(cell_barcodes)]
    
    stats = {
        'estimated_number_of_cells': len(truecellreads),
        'unique_reads_in_cells_mapped_to_gene': truecellreads['featureU'].sum(),
        'fraction_of_unique_reads_in_cells': round(truecellreads['featureU'].sum()/cellreadsdata['featureU'].sum(), 4),
        'mean_reads_per_cell': round(truecellreads['featureU'].mean(), 0),
        'median_reads_per_cell': round(truecellreads['featureU'].median(), 0),
        'umi_in_cells': truecellreads['nUMIunique'].sum(),
        'mean_umi_per_cell': round(truecellreads['nUMIunique'].mean(), 0),
        'median_umi_per_cell': round(truecellreads['nUMIunique'].median(), 0),
        'mean_gene_per_cell': round(truecellreads['nGenesUnique'].mean(), 0),
        'median_gene_per_cell': round(truecellreads['nGenesUnique'].median(), 0),
        'median_top100_umi_cell_mean': round(cb_umi_mean_top100['umi_count_mean'].median(), 3),
        'mean_top100_umi_cell_mean': round(cb_umi_mean_top100['umi_count_mean'].mean(), 3)
    }
    
    with open(os.path.join(outdir, "filtered_cells.summary.csv"), 'w') as f:
        for k, v in stats.items():
            f.write(f"{k},{v}\n")


def dbscan_filter(infile, outdir, maxumi, minumi, matrixdir, cellreads, library, eps, min_samples, n_jobs):
    os.makedirs(outdir, exist_ok=True)
    # Process data
    
    df = read_data(infile)
    filtered_df, subsb_umi_summary = filter_by_umi(df, maxumi, minumi)
    coord_df, cb_cluster, cluster_stats = perform_dbscan_parallel(filtered_df, eps, min_samples, n_jobs)
    
    # Save intermediate results
    cluster_stats.to_csv(os.path.join(outdir, "dbscan_cluster_distribution.csv"), index=False)
    cb_cluster.to_csv(os.path.join(outdir, "CB_cluster.txt"), sep='\t', index=False)
    subsb_umi_summary.to_csv(os.path.join(outdir, "spatial_umi_knee_plot.temp.txt"), sep='\t', index=False)
    
    # Generate plots and additional data
    cb_umi_mean_top100 = generate_plots(filtered_df, cb_cluster, subsb_umi_summary, outdir)
    cb_umi_mean_top100.to_csv(os.path.join(outdir, "cb_umi_mean_knee_plot.temp.txt"), sep='\t', index=False)

    # Filter matrix and generate final outputs
    filter_matrix(matrixdir, coord_df, library, outdir)
    generate_summary(cellreads, coord_df, cb_umi_mean_top100, outdir)


def parse_args():
    parser = argparse.ArgumentParser(
        description='Perform DBSCAN clustering for cell barcodes and filter matrix.',
        formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )
    parser.add_argument('-i', '--infile',
        metavar='FILE',
        type=str,
        required=True,
        help='Input coordinate file containing cb, sb, xcoord, ycoord, umi_count'
        )
    parser.add_argument('-o', '--outdir',
        metavar='PATH',
        type=str,
        required=True,
        help='Output directory for results'
        )
    parser.add_argument('-m', '--matrixdir',
        metavar='PATH',
        type=str,
        required=True,
        help='Directory containing 10X format matrix (raw_feature_bc_matrix)'
        )
    parser.add_argument('-cr', '--cellreads',
        metavar='FILE',
        type=str,
        required=True,
        help='Cell reads file from STARsolo output'
        )
    parser.add_argument('-l', '--library',
        metavar='STR',
        type=str,
        choices=['10X', 'leader_v1'],
        default='10X',
        help='Library type: 10X or leader_v1'
        )
    parser.add_argument('--maxumi',
        metavar='INT',
        type=int,
        default=5000,
        help='Maximum UMI count threshold for spatial barcodes filtering'
        )
    parser.add_argument('--minumi',
        metavar='INT',
        type=int,
        default=2,
        help='Minimum UMI count threshold for spatial barcodes filtering'
        )
    parser.add_argument('--eps',
        metavar='FLOAT',
        type=float,
        default=150.0,
        help='DBSCAN epsilon parameter (maximum distance between points)'
        )
    parser.add_argument('--min_samples',
        metavar='INT',
        type=int,
        default=6,
        help='DBSCAN min_samples parameter (minimum points to form cluster)'
        )
    parser.add_argument('--n_jobs',
        metavar='INT',
        type=int,
        default=-1,
        help='Number of CPU cores to use (-1 for all available)'
        )
    
    return parser.parse_args()


if __name__ == '__main__':
    args = parse_args()
    dbscan_filter(args.infile, args.outdir, args.maxumi, args.minumi, 
                  args.matrixdir, args.cellreads, args.library, 
                  args.eps, args.min_samples, args.n_jobs)
